packages/zero-pg/src/custom.ts

```diff
@@ -75,22 +78,168 @@ export class TransactionImpl<S extends Schema, TWrappedTransaction>
 
   constructor(
     dbTransaction: DBTransaction<TWrappedTransaction>,
-    _schema: S,
     token: string | undefined,
     clientID: string,
     mutationID: number,
+    mutate: SchemaCRUD<S>,
+    query: SchemaQuery<S>,
   ) {
     this.dbTransaction = dbTransaction;
     this.token = token;
     this.clientID = clientID;
     this.mutationID = mutationID;
+    this.mutate = mutate;
+    this.query = query;
+  }
+}
 
-    // TODO: implement both of these.
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    this.mutate = {} as any;
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    this.query = {} as any;
-    // this.mutate = makeSchemaCRUD(schema, dbTransaction);
-    // this.query = makeSchemaQuery(schema, dbTransaction);
+const dbTxSymbol = Symbol();
+type WithHiddenTx = {[dbTxSymbol]: DBTransaction<unknown>};
+
+export function makeSchemaCRUD<S extends Schema>(
+  schema: S,
+): (dbTransaction: DBTransaction<unknown>) => SchemaCRUD<S> {
+  const schemaCRUDs: Record<string, TableCRUD<TableSchema>> = {};
+  for (const tableSchema of Object.values(schema.tables)) {
+    schemaCRUDs[tableSchema.name] = makeTableCRUD(tableSchema);
+  }
+
+  /**
+   * For users with very large schemas it is expensive to re-create
+   * all the CRUD mutators for each transaction. Instead, we create
+   * them all once up-front and then bind them to the transaction
+   * as requested.
+   */
+  class CRUDHandler {
+    readonly #dbTransaction: DBTransaction<unknown>;
+    constructor(dbTransaction: DBTransaction<unknown>) {
+      this.#dbTransaction = dbTransaction;
+    }
+
+    get(target: Record<string, TableCRUD<TableSchema>>, prop: string) {
+      if (prop in target) {
+        return target[prop];
+      }
+
+      const txHolder: WithHiddenTx = {
+        [dbTxSymbol]: this.#dbTransaction,
+      };
+      target[prop] = Object.fromEntries(
+        Object.entries(schemaCRUDs[prop]).map(([name, method]) => [
+          name,
+          method.bind(txHolder),
+        ]),
+      ) as TableCRUD<TableSchema>;
+
+      return target[prop];
+    }
   }
+
+  return (dbTransaction: DBTransaction<unknown>) =>
+    new Proxy({}, new CRUDHandler(dbTransaction)) as SchemaCRUD<S>;
```

Comment:
Do we really need a Proxy?

For Replicache, the tx creation was hot and Proxies are not fast.